### PR TITLE
Fix ranch_server ETS table before match in get_listener_sups/0

### PR DESCRIPTION
**Pull request by Zalastax**
It was easy to make get_listener_sups/0 crash before this - probably when the ?TAB was not yet created.

I used to get this error:
```Erlang{badarg,
 [{ets,match,
   [ranch_server,
    {{listener_sup,
      '$1'},
     '$2'}],
   []},
  {ranch_server,
   get_listener_sups,
   0,
   [{file,
     ".../ranch/src/ranch_server.erl"},
    {line,105}]},
  {ranch,info,0,
   [{file,
     ".../ranch/src/ranch.erl"},
    {line,334}]},
...
```

### CODE REVIEW

**Comment by juhlig** [comment]
Can you please shed some light on how you manage to provoke that behavior, and provide the full error message?

Because, really, that should not be possible unless the `ranch` application is not running at all. The table is created in and is owned by the application's top supervisor, even before `ranch_server` itself is started, and remains even if `ranch_server` crashes.

However that may be, I think we really should go looking for the _cause_ of the issue if there is really any, instead of hushing up the symptoms. @essen, you agree?

**Comment by essen** [comment]
I'm fine with adding a try/catch as long as I understand why.

**Comment by Zalastax** [comment]
I'm not allowed to share the full error message. The rest of the trace just shows my own application code. However, @juhlig, your question provoked me to investigate a bit further. Turns out that [the error happens due to the table being modified while it's being iterated](https://stackoverflow.com/a/37953974/497116). This aligns well with how I triggered the error: my property based tests add and remove listeners rapidly while also checking the current status using `ranch:info/0`. I have a new patch that solves the issue in the way suggested on StackOverflow.

**Comment by essen** [comment]
Sounds good!

**Comment by juhlig** [comment]
Ok, in that case, there are more places where `ets:match` is used, eg `get_connections_sups/1,2`, and some more where `ets:match_delete` is used, eg `cleanup_listener_opts/1` and `cleanup_connections_sups/1`. As those would be equally affected, they should be fixed in the same manner.

**Comment by essen** [comment]
What's the error type, a function_clause? Badarg? Something else?

**Comment by Zalastax** [comment]
`error:badarg`, as caught in https://github.com/ninenines/ranch/commit/b42b87a7ef9e6f0f0128ed4acf9fc7b96dbf3eea. Bad design by OTP to report that error for this fault.

**Comment by essen** [comment]
Alright thanks.

You mentioned a property based test, could you provide it as well? We can add proper if necessary. Even if it's incomplete or doesn't work with proper we can probably adapt it.

**Comment by Zalastax** [comment]
Unfortunately I can't share that code :cry: But the test exercises the cowboy API by registering and deregistering on different ports. We verify the results via ranch info. It's quite some effort to set it all up so you would have to be quite dedicated to add it here.

**Comment by Zalastax** [comment]
Anything else you want me to add or is this deliverable?

**Comment by essen** [comment]
Well if you can write a test that'd help, to make sure we are catching everything. Otherwise please wait for us to process it.

**Comment by juhlig** [comment]
As I pointed out, there are more places in `ranch_server` where `match` (and `match_delete`, for that matter) is used, and which could IMO be fixed at the same time.

**Comment by juhlig** [comment]
Actually, the behavior leading to this PR is somewhat mysterious... I re-read the ets docs, specifically the part about [traversal](http://erlang.org/doc/man/ets.html#traversal). There it says that traversal is implicitly safe if the table is of type `ordered_set`, which the `ranch_server` table is. Further, it implies that entire traversals done within a single ETS function call are implicitly safe as well, which `ets:match/2` is...

**Comment by essen** [comment]
If there's a problem in OTP then it should be easy enough to reproduce it.

**Comment by juhlig** [comment]
@Zalastax can you confirm that `safe_fixtable` solves the problem? After re-reading that SO article, what is done there is an iteration over the table with `select/3` and `select/1` in chunks, not a full traversal as `match/2` (or `select/2`, for that matter) does, which according to the docs should be inherently safe ("done with a single ETS function call"), even if the table was not an `ordered_set`.

**Comment by juhlig** [comment]
@Zalastax I tried to reproduce the error, but couldn't. What I did was this:

* start 100 processes, each starting and stopping an own ranch_listener, 1000 times, with a random 0..10ms interval in between
* have another process call `ranch_server:get_listener_sups/0`, likewise 1000 times with a ~10ms interval; I didn't use `ranch:info/0` there because with the listeners coming and going like this, some of the `ranch_conns_sup_sup:which_children` calls would be likely to fail because it had just been turned off, but that would not be the error you describe and fix here

In short, this works just as expected, nothing bad happens. Tried with OTP 20, 21 and 22. So, please provide some more info on what you're really doing, and with what version of OTP and ranch. Better still, provide a test case that fails without the fix and does not fail with it. None of your confidential application-specific code is needed.

**Comment by Zalastax** [comment]
Sorry for wasting your time. I'm ashamed that I didn't perform this investigation in a better manner.

It turns out that proper is more parallel than we thought, resulting in `ranch:info()` sometimes being called before `application:ensure_started(ranch)` due to how we have written our tests. The whole call chain of `ranch:info()` is performed in the calling process so instead of getting a `noproc` as when interacting with a genserver, we reach `ets:match` but the ets table is not created yet since `application:ensure_started(ranch)` has not been called. I was able to fix the problem by moving `application:ensure_started(ranch)`.

**Comment by juhlig** [comment]
@Zalastax no problem :)
